csharp/src/Ice/IceSSL/TransceiverI.cs

```diff
@@ -512,278 +447,111 @@ private void finishAuthenticate()
-        string[] acceptableIssuers)
-    {
-        if (certs == null || certs.Count == 0)
-        {
-            return null;
-        }
-        else if (certs.Count == 1)
-        {
-            return certs[0];
-        }
-
-        // Use the first certificate that match the acceptable issuers.
-        if (acceptableIssuers != null && acceptableIssuers.Length > 0)
-        {
-            foreach (X509Certificate certificate in certs)
-            {
-                if (Array.IndexOf(acceptableIssuers, certificate.Issuer) != -1)
-                {
-                    return certificate;
-                }
-            }
-        }
-        return certs[0];
-    }
-
     private bool validationCallback(
         object sender,
         X509Certificate certificate,
-        X509Chain chainEngine,
+        X509Chain systemChain,
         SslPolicyErrors policyErrors)
     {
-        using var chain = new X509Chain(_instance.engine().useMachineContext());
-        try
+        int errors = (int)policyErrors;
+        int traceLevel = _instance.securityTraceLevel();
+        string traceCategory = _instance.securityTraceCategory();
+        Ice.Logger logger = _instance.logger();
+        bool checkCertName = _instance.engine().getCheckCertName();
+        string message = "";
+
+        X509Certificate2Collection caCerts = _instance.engine().caCerts();
+        X509ChainStatus[] chainStatus = [];
+        if (caCerts is null)
+        {
+            // If the CA certs are null, we use the provided X509Chain, which is built using the system CAs.
+            chainStatus = systemChain.ChainStatus;
+        }
+        else
         {
-            if (_instance.checkCRL() == 0)
+            using var chain = new X509Chain(_instance.engine().useMachineContext());
+            chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+            chain.ChainPolicy.DisableCertificateDownloads = true;
+            chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
+            foreach (X509Certificate2 cert in caCerts)
             {
-                chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                chain.ChainPolicy.CustomTrustStore.Add(cert);
             }
 
-            X509Certificate2Collection caCerts = _instance.engine().caCerts();
-            if (caCerts != null)
+            if (certificate is not null)
             {
-                // We need to set this flag to be able to use a certificate authority from the extra store.
-                chain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority;
-                foreach (X509Certificate2 cert in caCerts)
+                bool verified = chain.Build((X509Certificate2)certificate);
+                if (verified)
                 {
-                    chain.ChainPolicy.ExtraStore.Add(cert);
+                    errors ^= (int)SslPolicyErrors.RemoteCertificateChainErrors;
```

Comment:
Why the switch to exclusive or? How do we deal with the case where `policyErrors` already has something, specifically something that we're checking ourselves manually.

Reply:
so use `LoadFromPemFile` for `.pem` insensitive compare, and everything else for `Import` which handles pfx/der?

Note `LoadFromPemFile` is the only way to get multiple certificates imported.

Reply:
?

Reply:
replied to the wrong comment

Reply:
Reworked this so that it is not required. We now set the custom trust store in the auth options, and the chain passed to the verification callback is built using the custom trust store when the user configures a custom CA.